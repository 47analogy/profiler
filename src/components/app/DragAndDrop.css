.dragAndDropArea {
  isolation: isolate;
}

.dragAndDropMessageWrapper {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  align-items: stretch;
  justify-content: center;
  padding: 1em;
  background-color: rgba(255, 255, 255, 0.9);
  opacity: 0;
  pointer-events: none;
  transition: opacity 200ms;
}

.dragAndDropMessageWrapper.dragging {
  opacity: 1;
}

.dragAndDropMessage {
  display: flex;
  width: 100%;
  align-items: center;
  justify-content: center;
  border: 2px dashed #888;
  border-radius: 10px;
  color: #666;
  font-size: 30px;
  text-align: center;
}
