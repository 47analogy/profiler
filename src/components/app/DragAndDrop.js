/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';
import { retrieveProfileFromFile } from '../../actions/receive-profile';
import type { ConnectedProps } from '../../utils/connect';

import './DragAndDrop.css';

function _dragPreventDefault(event: DragEvent) {
  event.preventDefault();
}

type OwnProps = {|
  +className?: string,
  +children?: React.Node,
  +render?: React.Node => React.Node,
|};

type DispatchProps = {|
  +retrieveProfileFromFile: typeof retrieveProfileFromFile,
|};

type DragAndDropState = {
  isDragging: boolean,
};

type DragAndDropProps = ConnectedProps<OwnProps, {||}, DispatchProps>;

// TODO Add documentation

/**
 * Creates a target area to drop files on. A dropped file will be
 * loaded into the profiler.
 */
class DragAndDrop extends React.PureComponent<
  DragAndDropProps,
  DragAndDropState
> {
  state = {
    isDragging: false,
  };

  componentDidMount() {
    // Prevent dropping files on the document.
    document.addEventListener('drag', _dragPreventDefault, false);
    document.addEventListener('dragover', _dragPreventDefault, false);
    document.addEventListener('drop', _dragPreventDefault, false);
  }

  componentWillUnmount() {
    document.removeEventListener('drag', _dragPreventDefault, false);
    document.removeEventListener('dragover', _dragPreventDefault, false);
    document.removeEventListener('drop', _dragPreventDefault, false);
  }

  _startDragging = (event: Event) => {
    event.preventDefault();
    this.setState({ isDragging: true });
  };

  _stopDragging = (event: Event) => {
    event.preventDefault();
    this.setState({ isDragging: false });
  };

  _handleProfileDrop = (event: DragEvent) => {
    event.preventDefault();
    this.setState({ isDragging: false });

    if (!event.dataTransfer) {
      return;
    }

    const { files } = event.dataTransfer;
    if (files.length > 0) {
      this.props.retrieveProfileFromFile(files[0]);
    }
  };

  render() {
    const { className, children, render } = this.props;

    const message = (
      <div
        className={classNames(
          'dragAndDropMessageWrapper',
          this.state.isDragging ? 'dragging' : false
        )}
      >
        <div className="dragAndDropMessage">Drop a saved profile here</div>
      </div>
    );

    return (
      <>
        <div
          className={classNames(className, 'dragAndDropArea')}
          onDragEnter={this._startDragging}
          onDragExit={this._stopDragging}
          onDrop={this._handleProfileDrop}
        >
          {render ? render(message) : children}
        </div>
        {/* If we weren't provided a render prop, have the message div
          as a sibling to the area div above. The area div creates its
          own stacking context, so even if it contains children with
          high z-indexes, the message div will still appear on top
          when shown.*/
        render === undefined && message}
      </>
    );
  }
}

export default DragAndDrop;
