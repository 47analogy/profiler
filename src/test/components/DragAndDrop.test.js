/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import DragAndDrop from '../../components/app/DragAndDrop';

describe('app/DragAndDrop', () => {
  it('matches the snapshot', () => {
    const retrieveProfileFromFile = jest.fn();
    const { container } = render(
      <DragAndDrop retrieveProfileFromFile={retrieveProfileFromFile}>
        Target area here
      </DragAndDrop>
    );
    const [targetArea, message] = container.children;
    expect(targetArea).toMatchSnapshot();
    expect(message).toMatchSnapshot();
  });

  it('responds to dragging', () => {
    // It would be better if we could check that when dropping
    // something on the area we'd get a call to
    // `retrieveProfileFromFile`, but jsdom is not supporting
    // `dataTransfer`. See
    // https://github.com/testing-library/react-testing-library/issues/339
    // for more info.
    const retrieveProfileFromFile = jest.fn();
    const { container } = render(
      <DragAndDrop retrieveProfileFromFile={retrieveProfileFromFile}>
        Target area here
      </DragAndDrop>
    );
    const [targetArea, message] = container.children;
    expect(message.classList).not.toContain('dragging');

    fireEvent.dragEnter(targetArea);
    expect(message.classList).toContain('dragging');

    fireEvent.dragExit(targetArea);
    expect(message.classList).not.toContain('dragging');
  });
});
