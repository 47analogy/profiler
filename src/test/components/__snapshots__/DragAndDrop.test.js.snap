// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/DragAndDrop matches the snapshot 1`] = `
<div
  class="dragAndDropArea"
>
  Target area here
</div>
`;

exports[`app/DragAndDrop matches the snapshot 2`] = `
<div
  class="dragAndDropMessageWrapper"
>
  <div
    class="dragAndDropMessage"
  >
    Drop a saved profile here
  </div>
</div>
`;
